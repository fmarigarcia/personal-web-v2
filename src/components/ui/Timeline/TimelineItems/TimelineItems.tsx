import React from 'react';
import type { ExperienceItem } from '@data/experiences';
import { TimelinePosition } from '../TimelinePosition';
import { TimelineDot } from '../TimelineDot';
import { TimelineDuration } from '../TimelineDuration';

interface TimelineItemsProps {
    experiences: ExperienceItem[];
    selectedExp: ExperienceItem;
    onSelectExp: (exp: ExperienceItem) => void;
    onHoverExp: (expId: string | null) => void;
}

export const TimelineItems: React.FC<TimelineItemsProps> = ({
    experiences,
    selectedExp,
    onSelectExp,
    onHoverExp,
}) => {
    return (
        <div className="flex justify-between items-start relative z-10">
            {experiences.map((exp) => {
                const isSelected = selectedExp.id === exp.id;

                return (
                    <button
                        key={exp.id}
                        className="flex flex-col items-center group cursor-pointer min-w-0 max-w-40"
                        onClick={() => onSelectExp(exp)}
                        onMouseEnter={() => onHoverExp(exp.id)}
                        onMouseLeave={() => onHoverExp(null)}
                    >
                        <TimelinePosition
                            company={exp.company}
                            position={exp.position}
                            isSelected={isSelected}
                        />
                        <TimelineDot isSelected={isSelected} />
                        <TimelineDuration
                            duration={exp.duration}
                            isSelected={isSelected}
                        />
                    </button>
                );
            })}
        </div>
    );
};
