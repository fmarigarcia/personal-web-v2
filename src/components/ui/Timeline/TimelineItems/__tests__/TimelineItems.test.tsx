import { render, screen, fireEvent } from '@testing-library/react';
import type { ExperienceItem } from '../../../../../types/data';
import { TimelineItems } from '../TimelineItems';

// Mock usePlatform hook
jest.mock('../../../../../hooks/usePlatform', () => ({
    usePlatform: jest.fn(() => ({
        data: {
            isDesktop: true,
            isMobile: false,
            isTablet: false,
        },
    })),
}));

// Mock the sub-components
jest.mock('../../TimelinePosition', () => ({
    TimelinePosition: ({
        company,
        position,
        isSelected,
    }: {
        company: string;
        position: string;
        isSelected: boolean;
    }) => (
        <div data-testid="timeline-position">
            {company} - {position} - {isSelected.toString()}
        </div>
    ),
}));

jest.mock('../../TimelineDot', () => ({
    TimelineDot: ({ isSelected }: { isSelected: boolean }) => (
        <div data-testid="timeline-dot">{isSelected.toString()}</div>
    ),
}));

jest.mock('../../TimelineDuration', () => ({
    TimelineDuration: ({
        duration,
        isSelected,
    }: {
        duration: string;
        isSelected: boolean;
    }) => (
        <div data-testid="timeline-duration">
            {duration} - {isSelected.toString()}
        </div>
    ),
}));

const mockExperiences: ExperienceItem[] = [
    {
        id: 'exp1',
        company: 'Company A',
        position: 'position1',
        duration: 'duration1',
        descriptionKeys: ['desc1'],
        technologies: ['Tech1'],
    },
    {
        id: 'exp2',
        company: 'Company B',
        position: 'position2',
        duration: 'duration2',
        descriptionKeys: ['desc2'],
        technologies: ['Tech2'],
    },
];

describe('TimelineItems', () => {
    const defaultProps = {
        experiences: mockExperiences,
        selectedExp: mockExperiences[0],
        onSelectExp: jest.fn(),
        onHoverExp: jest.fn(),
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should render timeline items for all experiences', () => {
        render(<TimelineItems {...defaultProps} />);

        expect(screen.getAllByTestId('timeline-position')).toHaveLength(2);
        expect(screen.getAllByTestId('timeline-dot')).toHaveLength(2);
        expect(screen.getAllByTestId('timeline-duration')).toHaveLength(2);
    });

    it('should pass correct props to sub-components for selected item', () => {
        render(<TimelineItems {...defaultProps} />);

        // First item should be selected (isSelected = true)
        expect(
            screen.getByText('Company A - position1 - true')
        ).toBeInTheDocument();
        expect(screen.getByText('duration1 - true')).toBeInTheDocument();

        // Second item should not be selected (isSelected = false)
        expect(
            screen.getByText('Company B - position2 - false')
        ).toBeInTheDocument();
        expect(screen.getByText('duration2 - false')).toBeInTheDocument();
    });

    it('should call onSelectExp when item is clicked', () => {
        const onSelectExp = jest.fn();
        render(<TimelineItems {...defaultProps} onSelectExp={onSelectExp} />);

        const buttons = screen.getAllByRole('button');
        fireEvent.click(buttons[1]); // Click second item

        expect(onSelectExp).toHaveBeenCalledWith(mockExperiences[1]);
    });

    it('should call onHoverExp when item is hovered', () => {
        const onHoverExp = jest.fn();
        render(<TimelineItems {...defaultProps} onHoverExp={onHoverExp} />);

        const buttons = screen.getAllByRole('button');

        fireEvent.mouseEnter(buttons[0]);
        expect(onHoverExp).toHaveBeenCalledWith('exp1');

        fireEvent.mouseLeave(buttons[0]);
        expect(onHoverExp).toHaveBeenCalledWith(null);
    });

    it('should have correct button structure and classes', () => {
        const { container } = render(<TimelineItems {...defaultProps} />);

        const wrapper = container.firstChild as HTMLElement;
        expect(wrapper).toHaveClass(
            'flex',
            'justify-between',
            'items-start',
            'relative',
            'z-10'
        );

        const buttons = screen.getAllByRole('button');
        buttons.forEach((button) => {
            expect(button).toHaveClass(
                'flex',
                'flex-col',
                'items-center',
                'group',
                'cursor-pointer',
                'w-40'
            );
        });
    });

    it('should handle empty experiences array', () => {
        render(<TimelineItems {...defaultProps} experiences={[]} />);

        expect(
            screen.queryByTestId('timeline-position')
        ).not.toBeInTheDocument();
        expect(screen.queryByTestId('timeline-dot')).not.toBeInTheDocument();
        expect(
            screen.queryByTestId('timeline-duration')
        ).not.toBeInTheDocument();
    });

    it('should handle different selected experience', () => {
        render(
            <TimelineItems {...defaultProps} selectedExp={mockExperiences[1]} />
        );

        // First item should not be selected
        expect(
            screen.getByText('Company A - position1 - false')
        ).toBeInTheDocument();

        // Second item should be selected
        expect(
            screen.getByText('Company B - position2 - true')
        ).toBeInTheDocument();
    });

    it('should match snapshot', () => {
        const { container } = render(<TimelineItems {...defaultProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});
