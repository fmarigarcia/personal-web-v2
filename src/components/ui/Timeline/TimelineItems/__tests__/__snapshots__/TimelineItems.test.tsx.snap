// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineItems should match snapshot 1`] = `
<div
  class="flex justify-between items-start relative z-10"
>
  <button
    class="flex flex-col items-center group cursor-pointer w-40"
  >
    <div
      data-testid="timeline-position"
    >
      Company A
       - 
      position1
       - 
      true
    </div>
    <div
      data-testid="timeline-dot"
    >
      true
    </div>
    <div
      data-testid="timeline-duration"
    >
      duration1
       - 
      true
    </div>
  </button>
  <button
    class="flex flex-col items-center group cursor-pointer w-40"
  >
    <div
      data-testid="timeline-position"
    >
      Company B
       - 
      position2
       - 
      false
    </div>
    <div
      data-testid="timeline-dot"
    >
      false
    </div>
    <div
      data-testid="timeline-duration"
    >
      duration2
       - 
      false
    </div>
  </button>
</div>
`;
